### PR TITLE
Fix extract footnotes issue

### PR DESCRIPTION
**Pull request by rtonsing**

### CODE REVIEW

**Comment by rtonsing** [comment]
Might hold off for a bit - it is working, but it is putting a lot of empty lines in the footnotes list.

**Review by windymilla — APPROVED** [review]

File: ppcomp/ppcomp.py [hunk 1]
```diff
@@ -373,47 +373,41 @@ def extract_footnotes_pp(self):
         """Extract footnotes from a PP text file. Text is iterable. Returns the text as an iterable,
         without the footnotes, and footnotes as a list of (footnote string id, line number of the
         start of the footnote, list of strings comprising the footnote). fn_regexes is a list of
-        (regex, fn_type) that identify the beginning and end of a footnote. The fn_type is 1 when
-        a ] terminates it, or 2 when a new block terminates it.
+        (regex, fn_type) that identify the beginning and end of a footnote. The fn_type is 2 when
+        a ] terminates it, or 1 when a new block terminates it.
```
Comment (srjfoo) [inline ppcomp/ppcomp.py:377]: Bearing in mind that I'm not a PPer, when would a new block terminate a footnote, and not a closing square bracket?

Comment (rtonsing) [reply ppcomp/ppcomp.py:377]: Most of the time, actually. There are an annoying number of styles, but even the default Guiguts fixup results in:
`[1] This is a footnote.`

**Comment by rtonsing** [comment]
Still tweaking. At least I understand the code better now, such as putting out blank lines to maintain the line numbers.

File: ppcomp/ppcomp.py (hunk 1 above)
Comment (srjfoo) [reply ppcomp/ppcomp.py:377]: Ohh! Thanks. I was thinking as a formatter, not as a PPer. Silly me!

**Comment by rtonsing** [comment]
Seems to be working OK for 3 styles of text footnotes.

File: README.md [hunk 2]
```diff
@@ -97,15 +97,15 @@ Note that formatting of footnotes by processors varies widely, only a few variat
...
+By default, a few reasonable rules are applied to the HTML (See the definition of DEFAULT_TRANSFORM_CSS in the source code). However, it may be necessary to go further to reduce the amount of noise.
 
 Currently, a few CSS targets are supported:
 ```
 ::before  -- add content before the tag
 ::after   -- add content after the tag
 ```
 
-Some transformations are also supported:,
+Some transformations are also supported: attribute text
```
Comment (cpeel) [inline README.md:108]: `attribute text`?

File: README.md [hunk 3]
```diff
@@ -119,3 +119,55 @@ By default footnote anchors are expected to be surrounded by brackets. If it is
 ```
 --css '.fnanchor:before { content: "["; } .fnanchor:after { content: "]"; }'
 ```
+
+## Footnote extraction styles supported
+
+##### html: `class="footnote"`
+
+	<div class="footnote">
```
Comment (cpeel) [inline README.md:127]: Please use ` ``` ` syntax for codeblocks to match the rest of the file.

File: README.md [hunk 4]
```diff
@@ -119,3 +119,55 @@ By default footnote anchors are expected to be surrounded by brackets. If it is
 ```
 --css '.fnanchor:before { content: "["; } .fnanchor:after { content: "]"; }'
 ```
+
+## Footnote extraction styles supported
+
+##### html: `class="footnote"`
```
Comment (cpeel) [inline README.md:125]: These should either be a `###` to be one more than header above it, or they should be formatted differently -- but being H5s is just weird.

File: README.md [hunk 5]
```diff
@@ -119,3 +119,55 @@ By default footnote anchors are expected to be surrounded by brackets. If it is
...
+
+	<div class="footnote">
+	  <a id="Footnote_926" href="#FNanchor_926" class="label">[926]</a> Kidston (94), p. 250.
+	</div>
+
+##### Text output from formatting rounds (`"projectID"` prefix): `[Footnote #:` followed by 1 or more paragraphs, with end bracket, and possible `*[Footnote:` continuation blocks.
```
Comment (cpeel) [inline README.md:131]: These are really really long headers, can we make them shorter and move the content into normal text or something similar?

**Comment by srjfoo** [comment]
I think I've got a working copy of the fixes in my [ppwb/ppcomp](https://www.pgdp.org/~srjfoo/ppwb/ppcomp.php) sandbox, but I haven't had a chance to do any testing yet.

I do have some questions about the documentation, though. It looks like the "Usage" instructions are designed for local installations. That seems proper, but `ppwb` links to them, and doesn't give a separate run-down on the web options in the `ppwb` version. It doesn't look to me like all of the locally runnable options are available via `ppwb`, either, and I don't see a way to specify them. Am I missing something? (I tried to make sure I had the most recent version of `ppwb` running for the sandbox, though I may have missed something.)

**Comment by rtonsing** [comment]
Agreed, but ppcomp is written as a standalone command line utility, I feel it is up to the web page to present the description and options desired, and pass them to the utility. We can work on improving the ppwb project code. I've already requested an addition of an edit box to enter custom -css code.

**Comment by srjfoo** [comment]
> Agreed, but ppcomp is written as a standalone command line utility, I feel it is up to the web page to present the description and options desired, and pass them to the utility. We can work on improving the ppwb project code. I've already requested an addition of an edit box to enter custom -css code.

👍 

I agree that it should be up to the `ppwb` documentation to document options, but mentioned it here, since `ppwb` links to this documentation.

**Review by cpeel — APPROVED** [review]
`README.md` changes look good. As do high-level review of the code but I haven't tested the functionality myself.

**Comment by rtonsing** [comment]
Finally had time to do more testing. With charliehoward4dp's sample code:

1. It fails on srjfoo's test site. Is there a way I can access any error logs?
2. It passes when I run it locally, but there is still an issue I thought I had fixed.
3. I need to remove some debug code.

**Comment by srjfoo** [comment]
@rtonsing -- the log for the most recent run (r62797178317de) is
```
uploaded file: /data/htdocs/d/ppwb/r62797178317de/h35-epubpreamblefix.html
uploaded file: /data/htdocs/d/ppwb/r62797178317de/t41_fix3.txt
command: ./pyvenv-proxy.sh ./bin/ppcomp/ppcomp/ppcomp.py --ignore-case --extract-footnotes --txt-cleanup-type 'b' '/data/htdocs/d/ppwb/r62797178317de/h35-epubpreamblefix.html' '/data/htdocs/d/ppwb/r62797178317de/t41_fix3.txt'
error: Traceback (most recent call last):
  File "./bin/ppcomp/ppcomp/ppcomp.py", line 1147, in <module>
    main()
  File "./bin/ppcomp/ppcomp/ppcomp.py", line 1132, in main
    html_content, file1, file2 = compare.do_process()
  File "./bin/ppcomp/ppcomp/ppcomp.py", line 859, in do_process
    files[i].cleanup()  # perform cleanup for each type of file
  File "./bin/ppcomp/ppcomp/ppcomp.py", line 335, in cleanup
    self.extract_footnotes_pp()
  File "./bin/ppcomp/ppcomp/ppcomp.py", line 437, in extract_footnotes_pp
    with open('newfoot.txt', 'w', encoding='utf-8') as file:
PermissionError: [Errno 13] Permission denied: 'newfoot.txt'
```

**Comment by cpeel** [comment]
Ah, yes, are those file writes the remainder of debugging?

**Comment by rtonsing** [comment]
This fixes the issue reported by charliehoward4dp, at least. Thought I could get more footnote styles working, but broke the 1st one, so I reverted & will try again later.

**Pull request merged**